pkg/skaffold/deploy/component/kubernetes/component.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/access"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/deploy/label"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/debugging"
+	k8sloader "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/loader"
+	k8slogger "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/logger"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/portforward"
+	k8sstatus "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/status"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/loader"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/log"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/status"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/sync"
+)
+
+// For testing
+var (
+	NewAccessor    = newAccessor
+	NewDebugger    = newDebugger
+	NewImageLoader = newImageLoader
+	NewLogger      = newLogger
+	NewMonitor     = newMonitor
+	NewSyncer      = newSyncer
+)
+
+func newAccessor(cfg portforward.Config, cli *kubectl.CLI, podSelector kubernetes.PodSelector, labeller label.Config) access.Accessor {
+	if !cfg.PortForwardOptions().Enabled() {
+		return &access.NoopAccessor{}
+	}
+	return portforward.NewForwarderManager(cli, podSelector, labeller.RunIDSelector(), cfg.Mode(), cfg.PortForwardOptions(), cfg.PortForwardResources())
+}
+
+func newDebugger(mode config.RunMode, podSelector kubernetes.PodSelector) debug.Debugger {
+	if mode != config.RunModes.Debug {
+		return &debug.NoopDebugger{}
+	}
+
+	return debugging.NewContainerManager(podSelector)
+}
+
+func newImageLoader(cfg k8sloader.Config, cli *kubectl.CLI) loader.ImageLoader {
+	if cfg.LoadImages() {
+		return k8sloader.NewImageLoader(cfg.GetKubeContext(), cli)
+	}
+	return &loader.NoopImageLoader{}
+}
+
+func newLogger(config k8slogger.Config, cli *kubectl.CLI, podSelector kubernetes.PodSelector) log.Logger {
+	return k8slogger.NewLogAggregator(cli, podSelector, config)
+}
+
+func newMonitor(cfg k8sstatus.Config, labeller *label.DefaultLabeller) status.Monitor {
```

Comment:
for a multi-config project, this'll get instantiated multiple times. The `Monitor` is written so that all k8s deployers share a single instance of a `Monitor`. That was being enforced using the `map[sting]Monitor` in `monitor/provider.go`.

I think this is also required for others like `Accessor` and whichever had a similar map keyed on the `kubeContext` to maintain a single instance.

I think that'll cause a regression https://github.com/GoogleContainerTools/skaffold/issues/6074.

Reply:
also separate instances per `kubeContext` keeps this implementation open to Skaffold supporting multiple deployment clusters in the future if we ever do that.

Reply:
yep you're right. I went ahead and restored the kubecontext-keyed map as a component package-level variable, which will house the singletons for the `Accessor` and `Monitor`, and removed that logic from those constructors directly.

I also changed the logic in the `accessor.Start()/Stop()` methods to use `singleflight.Do` instead of the locks I had in place, which should give the behavior we want.